File: scripts/_test_with_xls.js
const { calculateIndex, inferIndexFromBalance } = require('../index-engine/calcIndex.js');
const PROVINCES = require('../index-mini/cloudfunctions/calcIndex/provinces-data.js');
const fs = require('fs');
const path = require('path');

const data = JSON.parse(fs.readFileSync(path.join(__dirname, '_xls_yearly.json'), 'utf8'));
const pc = PROVINCES['jilin'];
const { start_year, start_month, total_months, monthly_avg, principal_sum, yearly } = data;

// 新逻辑：当前基数 → 当下指数贯穿 → 反推历年基数
function endYearOf(sy, sm, tm) {
  let rem = tm, m = sm, y = sy;
  while (rem > 0) {
    const mty = 13 - m; const mm = Math.min(rem, mty > 0 ? mty : 12);
    rem -= mm;
    if (rem > 0) { y += 1; m = 1; }
  }
  return y;
}
function buildCurrentBase(sy, sm, tm, cur, sal) {
  const eY = endYearOf(sy, sm, tm);
  let by = eY - 1; while (by >= 1990 && !(sal[by] > 0)) by--;
  const bs = sal[by];
  if (!bs) return { error: 'no salary' };
  const ci = cur / bs;
  let rem = tm, y = sy, m = sm; const out = [];
  while (rem > 0) {
    const mty = 13 - m; const mm = Math.min(rem, mty > 0 ? mty : 12);
    const p = sal[y - 1];
    out.push({ year: y, months: mm, baseAvg: p ? p * ci : null });
    rem -= mm; y += 1; m = 1;
  }
  return { contribution: out, currentIndex: ci, baseYear: by, baseSalary: bs };
}

console.log('========== 模式A 金标准：逐年明细正向 ==========');
const rStd = calculateIndex({ provinceConfig: pc, contribution: yearly, granularity: 'A' });
console.log(`平均指数=${rStd.avgIndex.toFixed(4)}  估算余额(含息)=${rStd.accountBalance.toFixed(2)}  月数=${rStd.totalMonths}`);

console.log('\n========== 模式1 当前基数：当下指数贯穿反推 ==========');
const bCur = buildCurrentBase(start_year, start_month, total_months, monthly_avg, pc.avg_salary_history);
const rCur = calculateIndex({ provinceConfig: pc, contribution: bCur.contribution, granularity: 'A' });
console.log(`当下指数=${bCur.currentIndex.toFixed(4)}  输出指数=${rCur.avgIndex.toFixed(4)}  估算余额(含息)=${rCur.accountBalance.toFixed(2)}`);

console.log('\n========== 模式2 余额反推：用金标准含息余额 ==========');
const rInfStd = inferIndexFromBalance({ provinceConfig: pc, contribution: { startYear: start_year, startMonth: start_month, totalMonths: total_months }, granularity: 'C', knownBalance: rStd.accountBalance });
console.log('反推返回字段:', Object.keys(rInfStd));
console.log(`反推指数=${rInfStd.inferredIndex.toFixed(4)}  收敛=${rInfStd.converged}`);

console.log('\n========== 模式2b 余额反推：用本金和(不含息, 演示偏低) ==========');
const rInfPri = inferIndexFromBalance({ provinceConfig: pc, contribution: { startYear: start_year, startMonth: start_month, totalMonths: total_months }, granularity: 'C', knownBalance: principal_sum });
console.log(`反推指数=${rInfPri.inferredIndex.toFixed(4)}  收敛=${rInfPri.converged}`);

console.log('\n========== 模式3 两者都填：对照 ==========');
console.log(`正向(当前基数)=${rCur.avgIndex.toFixed(4)}  vs  反推(含息余额)=${rInfStd.inferredIndex.toFixed(4)}`);

const result = {
  province: 'jilin', start_year, start_month, total_months, monthly_avg, principal_sum,
  std: { avgIndex: +rStd.avgIndex.toFixed(4), balance: +rStd.accountBalance.toFixed(2) },
  current: { avgIndex: +rCur.avgIndex.toFixed(4), balance: +rCur.accountBalance.toFixed(2), currentIndex: +bCur.currentIndex.toFixed(4) },
  infer_from_std: +rInfStd.inferredIndex.toFixed(4),
  infer_from_principal: +rInfPri.inferredIndex.toFixed(4),
  yearly
};
fs.writeFileSync(path.join(__dirname, '_xls_result.json'), JSON.stringify(result, null, 2), 'utf8');
console.log('\n已保存 scripts/_xls_result.json');

File: scripts/_test_with_xls.py
# -*- coding: utf-8 -*-
import json
path = r'C:\Users\14041\Documents\12 廉政档案\个人缴费信息查询.xls'
import xlrd
wb = xlrd.open_workbook(path)
sh = wb.sheet_by_index(0)

rows = []
for r in range(1, sh.nrows):
    period = str(int(sh.cell_value(r, 0)))
    base = float(sh.cell_value(r, 3))
    acc = float(sh.cell_value(r, 4))
    if not period or base == 0:
        continue
    y = int(period[:4]); m = int(period[4:6])
    rows.append({'year': y, 'month': m, 'base': base, 'acc': acc})

rows.sort(key=lambda x: (x['year'], x['month']))
total_months = len(rows)
start_year = rows[0]['year']; start_month = rows[0]['month']
bases = [x['base'] for x in rows]
monthly_avg = sum(bases) / total_months
principal_sum = sum(x['acc'] for x in rows)

from collections import defaultdict
year_map = defaultdict(list)
for x in rows:
    year_map[x['year']].append(x['base'])
yearly = []
for y in sorted(year_map):
    bs = year_map[y]
    yearly.append({'year': y, 'months': len(bs), 'baseAvg': round(sum(bs)/len(bs), 2)})

out = {
  'province': 'jilin', 'start_year': start_year, 'start_month': start_month,
  'total_months': total_months, 'monthly_avg': round(monthly_avg, 2),
  'principal_sum': round(principal_sum, 2),
  'yearly': yearly
}
with open('scripts/_xls_yearly.json', 'w', encoding='utf-8') as f:
    json.dump(out, f, ensure_ascii=False, indent=2)
print('解析完成: 记录数=%d 起止=%d-%02d ~ %d 月均基数=%.2f 本金和=%.2f 年度数=%d' %
      (total_months, start_year, start_month, yearly[-1]['year'], monthly_avg, principal_sum, len(yearly)))
print('已保存 scripts/_xls_yearly.json')
